helper/schema/schema.go

```diff
@@ -1155,6 +1162,11 @@ func (m schemaMap) validateType(
 		ws, es = m.validatePrimitive(k, raw, schema, c)
 	}
 
+	if schema.ReadOnly {
+		es = append(es, fmt.Errorf(
+			"%q: This field is readonly and cannot be set by users", k))
+	}
+
```

Comment:
Actually - looking again - how does this ever pass when `ReadOnly` is true?

I'll recommend another unit test below to flush out the behavior.